src/System.CommandLine/Invocation/InvocationPipeline.cs

```diff
@@ -71,11 +92,27 @@ internal static int Invoke(ParseResult parseResult)
                 {
                     for (var i = 0; i < parseResult.NonexclusiveActions.Count; i++)
                     {
-                        parseResult.NonexclusiveActions[i].Invoke(parseResult);
+                        var action = parseResult.NonexclusiveActions[i];
+
+                        switch (action)
+                        {
+                            case SynchronousCliAction syncAction:
+                                syncAction.Invoke(parseResult);
+                                break;
+                            case AsynchronousCliAction _:
+                                throw new InvalidOperationException($"{nameof(AsynchronousCliAction)} called within non-async invocation.");
```

Comment:
I am big supporter of throwing in such case.  🥇 

My only suggestion would be to throw when the synchronous invocation starts and we detect that a command with async action is in the symbol tree (so when user has `n` sync commands and only one async, they are always going to get the exception and immediately switch to `InvokeAsync`)

Reply:
Good point. It might be though that this code is unreachable and better suited as a `Debug.Assert`. This method is only called currently from `ParseResult.Invoke` which does sync-over-async if it detects that there are any synchronous actions.

Reply:
I enclosed this in a `#if DEBUG` block, since this state is currently unreachable via the public API.